datalad/nonasyncrunner.py

```diff
@@ -0,0 +1,162 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Wrapper for command and function calls, allowing for dry runs and output handling
+
+"""
+
+import logging
+import os
+import queue
+import subprocess
+import threading
+import time
+from typing import Any
+
+
+logger = logging.getLogger("datalad.runner")
+
+STDOUT_FILENO = 1
+STDERR_FILENO = 2
+
+
+class ReaderThread(threading.Thread):
+    def __init__(self, file, a_queue, command):
+        super().__init__(daemon=True)
+        self.file = file
+        self.queue = a_queue
+        self.command = command
+        self.quit = False
+
+    def __str__(self):
+        return f"ReaderThread({self.file}, {self.queue}, {self.command})"
+
+    def request_exit(self):
+        """
+        Request the thread to exit. This is not guaranteed to
+        have any effect, because the thread might be waiting in
+        os.read() or queue.put().
+        """
+        self.quit = True
+
+    def run(self):
+        logger.debug("%s started", self)
+
+        while not self.quit:
+
+            data = os.read(self.file.fileno(), 1024)
+            if data == b"":
+                logger.debug("%s exiting (stream end)", self)
+                self.queue.put((self.file.fileno(), None, time.time()))
+                return
+
+            self.queue.put((self.file.fileno(), data, time.time()))
+
+
+def run_command(cmd,
+                protocol_class,
```

Comment:
```suggestion
                protocol,
```
to match the doc or adjust the doc?

Reply:
Good one. Let's adjust the doc

Reply:
sorry -- I should have provided more context and explicit desire right away:  we have it as a `protocol` for WitlessRunner as well and just document that it should be a class and not an instance.   It would be odd to use `protocol=` there and `protocol_class=` here. IMHO it would be better to stay consistent and just keep it as a `protocol` in the signature (and you could make it into `protocol_instance` or alike in the body).
FWIW  IIRC in typing should be smth like `: Type[WitlessProtocol]`.